File: packages/network_info_plus/network_info_plus/android/src/main/java/dev/fluttercommunity/plus/network_info/NetworkInfoPlusPlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package dev.fluttercommunity.plus.network_info;

import android.content.Context;
import android.net.wifi.WifiManager;
import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodChannel;

/** NetworkInfoPlusPlugin */
public class NetworkInfoPlusPlugin implements FlutterPlugin {

  private MethodChannel methodChannel;

  @Override
  public void onAttachedToEngine(FlutterPluginBinding binding) {
    setupChannels(binding.getBinaryMessenger(), binding.getApplicationContext());
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    teardownChannels();
  }

  private void setupChannels(BinaryMessenger messenger, Context context) {
    methodChannel = new MethodChannel(messenger, "dev.fluttercommunity.plus/network_info");
    WifiManager wifiManager =
        (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);

    NetworkInfo networkInfo = new NetworkInfo(wifiManager);

    NetworkInfoMethodChannelHandler methodChannelHandler =
        new NetworkInfoMethodChannelHandler(networkInfo);

    methodChannel.setMethodCallHandler(methodChannelHandler);
  }

  private void teardownChannels() {
    methodChannel.setMethodCallHandler(null);
    methodChannel = null;
  }
}

File: packages/network_info_plus/network_info_plus/android/src/main/kotlin/dev/fluttercommunity/plus/network_info/NetworkInfoPlusPlugin.kt
package dev.fluttercommunity.plus.network_info

import android.app.Activity
import android.content.Context
import android.net.wifi.WifiManager
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/** NetworkInfoPlusPlugin  */
class NetworkInfoPlusPlugin : FlutterPlugin, ActivityAware {
    private lateinit var methodChannel: MethodChannel
    private lateinit var methodChannelHandler: NetworkInfoMethodChannelHandler
    private lateinit var networkInfo: NetworkInfo

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        setupChannels(binding.binaryMessenger, binding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        setListenerActivity(binding.activity)
        binding.addRequestPermissionsResultListener(networkInfo)
    }

    override fun onDetachedFromActivity() {
        setListenerActivity(null)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    private fun setupChannels(messenger: BinaryMessenger, context: Context) {
        methodChannel = MethodChannel(messenger, CHANNEL)
        val wifiManager =
            context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
        networkInfo = NetworkInfo(wifiManager)
        methodChannelHandler = NetworkInfoMethodChannelHandler(networkInfo)
        methodChannel.setMethodCallHandler(methodChannelHandler)
    }

    private fun setListenerActivity(activity: Activity?) {
        if (::methodChannelHandler.isInitialized) {
            methodChannelHandler.activity = activity
        }
    }

    companion object {
        private const val CHANNEL = "dev.fluttercommunity.plus/network_info"
    }
}
